Convert use TestUtil components for functional tests

cacheung opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Revisit functional tests to use testUtils JsonAsserts APIs
Remove flatten usages in the functional tests
Remove utils which has been moved to and available on testUtils

## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

cacheung committed:
Functional Test update draft

Functional Test update draft

# Conflicts:
#	code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/util/IdentityFunctionalTestUtil.java

cacheung committed:
update IdentityPublicAPITest

update IdentityPublicAPITest

cacheung committed:
Update Functional tests, and remove new line for unit tests

Update Functional tests, and remove new line for unit tests

cacheung committed:
Fix spotlessCheck

Fix spotlessCheck

cacheung committed:
update testUtil library version and remove common library in Functional test

update testUtil library version and remove common library in Functional test

cacheung committed:
update functional test

update functional test
Using local testHelper for now due to a bug in testUtil, will revert while the bug is fixed.

cacheung committed:
Use TestHelper from the TestUtils

Use TestHelper from the TestUtils

cacheung committed:
fix format

fix format

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityAdIdTest.java:
```diff
@@ -12,7 +12,8 @@
 import static com.adobe.marketing.mobile.edge.identity.util.IdentityFunctionalTestUtil.*;
-import static com.adobe.marketing.mobile.edge.identity.util.TestHelper.*;
+import static com.adobe.marketing.mobile.util.NodeConfig.Scope.Subtree;
+import static com.adobe.marketing.mobile.util.TestHelper.*;
```
Can we replace the wildcard `*` imports with concrete imports? For quick fix you can use Android Studio built-in tool:
<img width="563" alt="Screenshot 2024-06-05 at 5 47 32 PM" src="https://github.com/adobe/aepsdk-edgeidentity-android/assets/95260439/dff61edf-6751-48e0-9cb3-d7f5fbbabd40">

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityAdIdTest.java:
```diff
@@ -337,37 +343,102 @@ private void verifyDispatchedEvents(boolean isGenericIdentityEventAdIdEvent, Str
-			Map<String, String> consentDataMap = flattenMap(dispatchedConsentEvents.get(0).getEventData());
-			assertEquals("GAID", consentDataMap.get("consents.adID.idType"));
-			assertEquals(expectedConsentValue, consentDataMap.get("consents.adID.val"));
+			Map<String, Object> consentDataMap = dispatchedConsentEvents.get(0).getEventData();
```
What do you think of slightly modifying this for conciseness?:
1. moving `expected` defn to the context right before it is used
2. replacing definition of single use `consentDataMap` with direct usage
3. Switching `assertTypeMatch` to `assertExactMatch` and removing `ValueExactMatch` path option

```java
if (!StringUtils.isNullOrEmpty(expectedConsentValue)) {
  String expected =
  "{" +
  "  \"consents\": {" +
  "    \"adID\": {" +
  "      \"idType\": \"GAID\"," +
  "      \"val\": \"" +
  expectedConsentValue +
  "\"" +
  "    }" +
  "  }" +
  "}";
  JSONAsserts.assertExactMatch(
    expected,
    dispatchedConsentEvents.get(0).getEventData(),
    new CollectionEqualCount(Subtree)
  );
}
```

Also should we apply the more strict `CollectionEqualCount(Subtree)` condition on the `consents` JSON? there is no flat map count requirement in the previous logic, so this would be more restrictive

cacheung replied:
Do you mean remove new CollectionEqualCount(Subtree) here?

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityAdIdTest.java:
```diff
@@ -337,37 +343,102 @@ private void verifyDispatchedEvents(boolean isGenericIdentityEventAdIdEvent, Str
 	 */
-	private void verifyFlatIdentityMap(
-		@NonNull final Map<String, String> flatIdentityMap,
+	private void verifyIdentityMap(
```
What do you think about simplifying the logic to something like:
```java
  private void verifyIdentityMap(
    @NonNull final Map<String, Object> identityMap,
    @Nullable final String expectedAdId
) {
    String baseIdentityMap = "{" +
        "\"identityMap\": {" +
        "    \"ECID\": [" +
        "        {" +
        "            \"id\": \"primaryECID\"," +
        "            \"authenticatedState\": \"ambiguous\"," +
        "            \"primary\": false" +
        "        }" +
        "    ]" +
        "%s" + // Placeholder for optional GAID
        "  }" +
        "}";

    String gaidTemplate = "," + // Leading comma required since GAID will be another property at the identityMap level
        "    \"GAID\": [" +
        "        {" +
        "            \"id\": \"" + expectedAdId + "\"," +
        "            \"authenticatedState\": \"ambiguous\"," +
        "            \"primary\": false" +
        "        }" +
        "    ]";

    String gaidProperty = expectedAdId != null ? gaidTemplate : "";
    String expectedJson = String.format(baseIdentityMap, gaidProperty);

    JSONAsserts.assertEquals(expectedJson, identityMap);
}
```

addb replied:
+1

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityBootUpTest.java:
```diff
@@ -51,20 +56,60 @@ public void testOnBootUp_LoadsAllIdentitiesFromPreference() throws Exception {
-		assertEquals("JohnDoe", xdmSharedState.get("identityMap.UserId[0].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
Should we more closely align this assertion to the original test case? something like:
```java
String expected = "{" +
  "\"identityMap\": {" +
  "    \"ECID\": [" +
  "        {" +
  "            \"id\": \"primaryECID\"" +
  "        }," +
  "        {" +
  "            \"id\": \"secondaryECID\"" +
  "        }" +
  "    ]," +
  "    \"Email\": [" +
  "        {" +
  "            \"id\": \"example@email.com\"" +
  "        }" +
  "    ]," +
  "    \"UserId\": [" +
  "        {" +
  "            \"id\": \"JohnDoe\"" +
  "        }" +
  "    ]" +
  "  }" +
  "}";
assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(12, Subtree), // 3 for ECID and 3 for secondaryECID + 6
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityBootUpTest.java:
```diff
@@ -51,20 +56,60 @@ public void testOnBootUp_LoadsAllIdentitiesFromPreference() throws Exception {
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(12, persistedMap.size()); // 3 for ECID and 3 for secondaryECID + 6
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
+		assertExactMatch(expected, persistedMap, new ElementCount(12, Subtree));
```
Should we update to use a generic object for `expected` to align with the original test case logic?:
```java
assertExactMatch("{}", persistedMap, new ElementCount(12, Subtree));
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityBootUpTest.java:
```diff
@@ -82,11 +127,36 @@ public void testOnBootUp_LoadsAllIdentitiesFromPreference_ignoresEmptyItems() th
-		assertEquals("primaryECID", xdmSharedState.get("identityMap.ECID[0].id"));
-		assertEquals("JohnDoe", xdmSharedState.get("identityMap.UserId[0].id"));
-		assertNull(xdmSharedState.get("identityMap.UserId[1].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
```
Should we more closely align this assertion to the original test case? something like:

```java
String expected = "{" +
  "\"identityMap\": {" +
  "    \"ECID\": [" +
  "        {" +
  "            \"id\": \"primaryECID\"" +
  "        }" +
  "    ]," +
  "    \"UserId\": [" +
  "        {" +
  "            \"id\": \"JohnDoe\"" +
  "        }" +
  "    ]" +
  "  }" +
  "}";
assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(6, Subtree), // 3 for ECID and 3 UserId JohnDoe
  new KeyMustBeAbsent("identityMap.UserId[1].id")
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -12,17 +12,25 @@
 import static com.adobe.marketing.mobile.edge.identity.util.IdentityFunctionalTestUtil.*;
-import static com.adobe.marketing.mobile.edge.identity.util.TestHelper.*;
+import static com.adobe.marketing.mobile.util.NodeConfig.Scope.Subtree;
+import static com.adobe.marketing.mobile.util.TestHelper.*;
```
Same thing here, can we replace the wildcard `*` imports with concrete imports?

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -56,12 +64,14 @@ public void testRegisterExtension_withClass() throws InterruptedException {
-			sharedStateMap.get("extensions.com.adobe.edge.identity.version")
+		Map<String, Object> sharedStateMap = getSharedStateFor(IdentityTestConstants.SharedStateName.EVENT_HUB, 5000);
+
+		String expected = "{\n" + "  \"version\": \"" + IdentityConstants.EXTENSION_VERSION + "\"\n" + "}";
```
I think the original test case is testing a different version:
```java
String expected = "{ " +
    "  \"extensions\": { " +
    "    \"com.adobe.edge.identity\": { " +
    "      \"version\": \"" + IdentityConstants.EXTENSION_VERSION + "\" " +
    "    } " +
    "  } " +
    "}";
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -56,12 +64,14 @@ public void testRegisterExtension_withClass() throws InterruptedException {
+		JSONAsserts.assertExactMatch(
+			expected,
+			sharedStateMap,
+			new ValueExactMatch("extensions.com.adobe.edge.identity.version")
```
Then you can remove this `ValueExactMatch` path option

timkimadobe replied:
Can we remove the `ValueExactMatch` from here? it should be covered by the `assertExactMatch`

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -80,22 +90,57 @@ public void testUpdateIdentitiesAPI() throws Exception {
-		assertEquals("ambiguous", xdmSharedState.get("identityMap.ECID[0].authenticatedState"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:
```java
String expected = "{" +
  "  \"identityMap\": {" +
  "    \"ECID\": [" +
  "      {" +
  "        \"authenticatedState\": \"ambiguous\"" +
  "      }" +
  "    ]," +
  "    \"Email\": [" +
  "      {" +
  "        \"id\": \"example@email.com\"," +
  "        \"primary\": true" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(6, Subtree), // 3 for ECID and 3 for Email
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -80,22 +90,57 @@ public void testUpdateIdentitiesAPI() throws Exception {
-		assertEquals("example@email.com", persistedMap.get("identityMap.Email[0].id"));
-		assertEquals("true", persistedMap.get("identityMap.Email[0].primary"));
-		assertEquals("ambiguous", persistedMap.get("identityMap.ECID[0].authenticatedState"));
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this a bit and also inline the `persistedJson` conversion to `JSONObject`:
```java
JSONAsserts.assertExactMatch(
  expected,
  new JSONObject(persistedJson),
  new ElementCount(6, Subtree), // 3 for ECID and 3 for Email
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -111,9 +156,27 @@ public void testUpdateAPI_nullData() throws Exception {
-		assertNotNull(xdmSharedState.get("identityMap.ECID[0].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected =
  "{" +
  "  \"identityMap\": {" +
  "    \"ECID\": [" +
  "    {" +
  "        \"id\": \"STRING_TYPE\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertTypeMatch(
  expected,
  xdmSharedState,
  new ElementCount(3, Subtree) // 3 for ECID still exists
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -129,9 +192,28 @@ public void testUpdateAPI_emptyData() throws Exception {
-		assertNotNull(xdmSharedState.get("identityMap.ECID[0].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected =
  "{" +
  "  \"identityMap\": {" +
  "    \"ECID\": [" +
  "    {" +
  "        \"id\": \"STRING_TYPE\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertTypeMatch(
  expected,
  xdmSharedState,
  new ElementCount(3, Subtree) // 3 for ECID still exists
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -149,22 +231,56 @@ public void testUpdateAPI_shouldReplaceExistingIdentities() throws Exception {
-		assertEquals("example@email.com", xdmSharedState.get("identityMap.Email[0].id"));
-		assertEquals("false", xdmSharedState.get("identityMap.Email[0].primary"));
-		assertEquals("loggedOut", xdmSharedState.get("identityMap.Email[0].authenticatedState"));
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected = "{" +
  "  \"identityMap\": {" +
  "    \"Email\": [" +
  "      {" +
  "        \"authenticatedState\": \"loggedOut\"," +
  "        \"id\": \"example@email.com\"," +
  "        \"primary\": false" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(6, Subtree) // 3 for ECID and 3 for Email
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -149,22 +231,56 @@ public void testUpdateAPI_shouldReplaceExistingIdentities() throws Exception {
-		assertEquals("example@email.com", persistedMap.get("identityMap.Email[0].id"));
-		assertEquals("false", persistedMap.get("identityMap.Email[0].primary"));
-		assertEquals("loggedOut", persistedMap.get("identityMap.Email[0].authenticatedState"));
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this a bit and also inline the `persistedJson` conversion to `JSONObject`:
```java
JSONAsserts.assertExactMatch(
  expected,
  new JSONObject(persistedJson),
  new ElementCount(6, Subtree) // 3 for ECID and 3 for Email
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -182,18 +298,42 @@ public void testUpdateAPI_withReservedNamespaces() throws Exception {
-		assertNotEquals("newECID", xdmSharedState.get("identityMap.ECID[0].id")); // ECID doesn't get replaced by API
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected =
  "{" +
  "  \"identityMap\": {" +
  "    \"ECID\": [" +
  "      {" +
  "        \"id\": \"newECID\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(3, Subtree), // 3 for ECID
  new ValueNotEqual("identityMap.ECID[0].id") // ECID doesn't get replaced by API
);

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -182,18 +298,42 @@ public void testUpdateAPI_withReservedNamespaces() throws Exception {
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(3, persistedMap.size()); // 3 for ECID
-		assertNotEquals("newECID", persistedMap.get("identityMap.ECID[0].id")); // ECID doesn't get replaced by API
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this a bit and also inline the `persistedJson` conversion to `JSONObject`:
```java
JSONAsserts.assertExactMatch(
  expected,
  new JSONObject(persistedJson),
  new ElementCount(3, Subtree), // 3 for ECID
  new ValueNotEqual("identityMap.ECID[0].id") // ECID doesn't get replaced by API
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -210,24 +350,77 @@ public void testUpdateAPI_multipleNamespaceMap() throws Exception {
-		assertEquals("secondary@email.com", xdmSharedState.get("identityMap.Email[1].id"));
-		assertEquals("zzzyyyxxx", xdmSharedState.get("identityMap.UserId[0].id"));
-		assertEquals("John Doe", xdmSharedState.get("identityMap.UserName[0].id"));
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java

String expected = "{" +
  "  \"identityMap\": {" +
  "    \"Email\": [" +
  "      {" +
  "        \"id\": \"primary@email.com\"" +
  "      }," +
  "      {" +
  "        \"id\": \"secondary@email.com\"" +
  "      }" +
  "    ]," +
  "    \"UserId\": [" +
  "      {" +
  "        \"id\": \"zzzyyyxxx\"" +
  "      }" +
  "    ]," +
  "    \"UserName\": [" +
  "      {" +
  "        \"id\": \"John Doe\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(15, Subtree) // 3 for ECID + 12 for new identities
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -210,24 +350,77 @@ public void testUpdateAPI_multipleNamespaceMap() throws Exception {
-		assertEquals("secondary@email.com", persistedMap.get("identityMap.Email[1].id"));
-		assertEquals("zzzyyyxxx", persistedMap.get("identityMap.UserId[0].id"));
-		assertEquals("John Doe", persistedMap.get("identityMap.UserName[0].id"));
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this a bit and also inline the `persistedJson` conversion to `JSONObject`:

```java

JSONAsserts.assertExactMatch(
  expected,
  new JSONObject(persistedJson),
  new ElementCount(15, Subtree) // 3 for ECID + 12 for new identities
);

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -242,20 +435,56 @@ public void testUpdateAPI_caseSensitiveNamespacesForCustomIdentifiers() throws E
-		assertEquals(9, xdmSharedState.size()); // 3 for ECID + 6 for new identities
-		assertEquals("primary@email.com", xdmSharedState.get("identityMap.Email[0].id"));
-		assertEquals("secondary@email.com", xdmSharedState.get("identityMap.email[0].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected = "{" +
  "  \"identityMap\": {" +
  "    \"Email\": [" +
  "      {" +
  "        \"id\": \"primary@email.com\"" +
  "      }" +
  "    ]," +
  "    \"email\": [" +
  "      {" +
  "        \"id\": \"secondary@email.com\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(9, Subtree) // 3 for ECID + 6 for new identities
);

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -242,20 +435,56 @@ public void testUpdateAPI_caseSensitiveNamespacesForCustomIdentifiers() throws E
-		assertEquals(9, persistedMap.size()); // 3 for ECID + 6 for new identities
-		assertEquals("primary@email.com", persistedMap.get("identityMap.Email[0].id"));
-		assertEquals("secondary@email.com", persistedMap.get("identityMap.email[0].id"));
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this a bit and also inline the `persistedJson` conversion to `JSONObject`:

```java

JSONAsserts.assertExactMatch(
  expected,
  new JSONObject(persistedJson),
  new ElementCount(9, Subtree) // 3 for ECID + 6 for new identities
);

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -406,25 +636,77 @@ public void testRemoveIdentity() throws Exception {
-		Map<String, String> xdmSharedState = flattenMap(getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000));
-		assertEquals(6, xdmSharedState.size()); // 3 for ECID + 3 for Email secondary
-		assertEquals("secondary@email.com", xdmSharedState.get("identityMap.Email[0].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected = "{" +
  "  \"identityMap\": {" +
  "    \"Email\": [" +
  "      {" +
  "        \"id\": \"secondary@email.com\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertTypeMatch(
  expected,
  xdmSharedState,
  new ElementCount(6, Subtree) // 3 for ECID + 3 for Email secondary
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -406,25 +636,77 @@ public void testRemoveIdentity() throws Exception {
-		assertEquals(3, xdmSharedState.size()); // 3 for ECID
+		xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expectedWithoutEmail =
```
I think we can simplify this to:

```java
// 3 for ECID
JSONAsserts.assertExactMatch("{}", xdmSharedState, new ElementCount(3, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -406,25 +636,77 @@ public void testRemoveIdentity() throws Exception {
 		);
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(3, persistedMap.size()); // 3 for ECID
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this to:

```java
// 3 for ECID
JSONAsserts.assertExactMatch("{}", new JSONObject(persistedJson), new ElementCount(3, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -437,16 +719,38 @@ public void testRemoveIdentity_nonExistentNamespace() throws Exception {
-		Map<String, String> xdmSharedState = flattenMap(getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000));
-		assertEquals(3, xdmSharedState.size()); // 3 for ECID
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+		String expected =
```
I think we can simplify this to:

```java
// 3 for ECID
JSONAsserts.assertExactMatch("{}", xdmSharedState, new ElementCount(3, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -437,16 +719,38 @@ public void testRemoveIdentity_nonExistentNamespace() throws Exception {
 		);
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(3, persistedMap.size()); // 3 for ECID
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this to:

```java
// 3 for ECID
JSONAsserts.assertExactMatch("{}", new JSONObject(persistedJson), new ElementCount(3, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -463,16 +767,46 @@ public void testRemoveIdentity_nameSpaceCaseSensitive() throws Exception {
-		assertEquals(6, xdmSharedState.size()); // 3 for ECID +  3 for  Email
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
I think we can simplify this to:

```java
// 3 for ECID + 3 for Email
JSONAsserts.assertExactMatch("{}", xdmSharedState, new ElementCount(6, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -463,16 +767,46 @@ public void testRemoveIdentity_nameSpaceCaseSensitive() throws Exception {
 		);
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(6, persistedMap.size()); // 3 for ECID +  3 for  Email
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this to:

```java
// 3 for ECID + 3 for Email
JSONAsserts.assertExactMatch("{}", new JSONObject(persistedJson), new ElementCount(6, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -489,16 +823,45 @@ public void testRemoveIdentity_nonExistentItem() throws Exception {
-
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
I think we can simplify this to:

```java
// 3 for ECID + 3 for Email
JSONAsserts.assertExactMatch("{}", xdmSharedState, new ElementCount(6, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -489,16 +823,45 @@ public void testRemoveIdentity_nonExistentItem() throws Exception {
 		);
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(6, persistedMap.size()); // 3 for ECID +  3 for  Email
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this to:

```java
// 3 for ECID + 3 for Email
JSONAsserts.assertExactMatch("{}", new JSONObject(persistedJson), new ElementCount(6, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -514,15 +877,37 @@ public void testRemoveIdentity_doesNotRemoveECID() throws Exception {
-		assertEquals(3, xdmSharedState.size()); // 3 for ECID that still exists
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
I think we can simplify this to:

```java
// 3 for ECID that still exists
JSONAsserts.assertExactMatch("{}", xdmSharedState, new ElementCount(3, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityPublicAPITest.java:
```diff
@@ -514,15 +877,37 @@ public void testRemoveIdentity_doesNotRemoveECID() throws Exception {
 		);
-		Map<String, String> persistedMap = flattenMap(JSONUtils.toMap(new JSONObject(persistedJson)));
-		assertEquals(3, persistedMap.size()); // 3 for ECID that still exists
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this to:

```java
// 3 for ECID that still exists
JSONAsserts.assertExactMatch("{}", new JSONObject(persistedJson), new ElementCount(3, Subtree));

```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityResetHandlingTest.java:
```diff
@@ -73,17 +78,44 @@ public void testReset_ClearsAllIDAndDispatchesResetComplete() throws Exception {
-		assertEquals(newECID, xdmSharedState.get("identityMap.ECID[0].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected =
  "{" +
  "  \"identityMap\": {" +
  "    \"ECID\": [" +
  "      {" +
  "        \"id\": \"" +
  newECID +
  "\"" +
  "      }" +
  "    ]" +
  "  }" +
  "}";

assertExactMatch(
  expected,
  xdmSharedState,
  new ElementCount(3, Subtree) // 3 for ECID still exists
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityResetHandlingTest.java:
```diff
@@ -73,17 +78,44 @@ public void testReset_ClearsAllIDAndDispatchesResetComplete() throws Exception {
-		assertEquals(3, persistedMap.size()); // 3 for ECID
-		assertEquals(newECID, persistedMap.get("identityMap.ECID[0].id"));
+
+		Map<String, Object> persistedMap = JSONUtils.toMap(new JSONObject(persistedJson));
```
I think we can simplify this to:

```java
assertExactMatch(
  expected,
  new JSONObject(persistedJson),
  new ElementCount(3, Subtree) // 3 for ECID still exists
);
```

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/util/IdentityFunctionalTestUtil.java:
```diff
@@ -351,10 +286,27 @@ public static void verifyPrimaryECIDNotNull() throws InterruptedException {
-			getXDMSharedStateFor(IdentityTestConstants.EXTENSION_NAME, 1000)
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityTestConstants.EXTENSION_NAME, 1000);
+
+		String expected =
```
To align better with the original test case logic I think this can be simplified to:

```java
String expected =
  "{" +
  "\"identityMap\": {" +
  "    \"ECID\": [" +
  "        {" +
  "            \"id\": \"STRING_TYPE\"" +
  "        }" +
  "    ]" +
  "  }" +
  "}";

JSONAsserts.assertTypeMatch(
  expected,
  xdmSharedState,
  new CollectionEqualCount(Subtree)
);
```
Can we also do a static import for `Subtree`?

cacheung replied:
Removed new CollectionEqualCount, we don't need it here.

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/util/IdentityFunctionalTestUtil.java:
```diff
@@ -386,17 +350,32 @@ public static void verifyPrimaryECID(final String primaryECID) throws Exception
-		assertEquals(secondaryECID, xdmSharedState.get("identityMap.ECID[1].id"));
+		Map<String, Object> xdmSharedState = getXDMSharedStateFor(IdentityTestConstants.EXTENSION_NAME, 1000);
+
+		if (secondaryECID == null) {
```
is `null` check required here? can this logic be simplified to:
```java
String expected =
    "{" +
    "  \"identityMap\": {" +
    "    \"ECID\": [" +
    "      {}," +  // Empty first ECID object
    "      {" +
    "        \"id\": \"" + secondaryECID + "\"" +
    "      }" +
    "    ]" +
    "  }" +
    "}";
JSONAsserts.assertExactMatch(expected, xdmSharedState);
```

cacheung replied:
They have few cases in IdentityECIDHandlingTest set verifySecondaryECID to null.  We need to cover for these cases.

timkimadobe replied:
Sorry, thank you for pointing this out! I misunderstood the intent of this logic. This portion looks great

Could we update the JSON string in this if block to (removes all the extra `+`):
```java
// Validates "ECID" array has a single object element, with no restriction on what is inside that element
String json = "{\"identityMap\": {\"ECID\": [{}]}}";
```

timkimadobe commented on code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityPropertiesTests.java:
```diff
@@ -136,7 +138,7 @@ public void test_toXDMData_OnlySecondaryECID() {
 		// test and verify, can't have secondary ECID without primary ECID
 		Map<String, Object> xdmMap = props.toXDMData(false);
-		JSONAsserts.assertEquals("{}", xdmMap);
+		JSONAsserts.assertExactMatch("{}", xdmMap, new ElementCount(0, Subtree));
```
I think the original `JSONAsserts.assertEquals("{}", xdmMap);` is ok here

timkimadobe reviewed (commented):
For the `new JSONObject(persistedJson)` i think we can just use `persistedJson` directly, since JSON comparison APIs will convert `String`s (forgot about that until I saw you use it in the later files)

addb commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityAdIdTest.java:
```diff
@@ -337,37 +343,102 @@ private void verifyDispatchedEvents(boolean isGenericIdentityEventAdIdEvent, Str
+					"identityMap.GAID[0].primary",
+					"identityMap.GAID[0].id",
+					"identityMap.GAID[0].authenticatedState, identityMap.ECID[0].primary, identityMap.ECID[0].id, identityMap.ECID[0].authenticatedState"
+				),
```
last three paths are one string which is not expected I guess.

```
new ValueExactMatch(
 					"identityMap.GAID[0].primary",
 					"identityMap.GAID[0].id",
 					"identityMap.GAID[0].authenticatedState",
 					"identityMap.ECID[0].primary, identityMap.ECID[0].id",                                          "identityMap.ECID[0].authenticatedState"
 				)
```

cacheung committed:
Apply optimize import for files

Apply optimize import for files

timkimadobe commented on code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityMapTests.java:
```diff
@@ -51,9 +53,9 @@ public void test_AddItem() {
-				Collections.singletonList("identityMap.location[0]"),
 				false,
-				NodeConfig.Scope.Subtree
+				NodeConfig.Scope.Subtree,
```
Can we do a static import of the `Subtree` option for this file?

timkimadobe commented on code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityMapTests.java:
```diff
@@ -51,9 +53,9 @@ public void test_AddItem() {
 				false,
-				NodeConfig.Scope.Subtree
+				NodeConfig.Scope.Subtree,
+				Collections.singletonList("identityMap.location[0]")
```
This one you should be able to remove the `Collections.singletonList()` wrapper and just use the path string directly: `"identityMap.location[0]"`

timkimadobe commented on code/edgeidentity/src/test/java/com/adobe/marketing/mobile/edge/identity/IdentityMapTests.java:
```diff
@@ -314,34 +316,27 @@ public void test_remove_nonexistentNamespaceAndItems() {
 
-		// verify
-		JSONAsserts.assertEquals(jsonStr, map.asXDMMap(false));
+		// verify  TODO
```
Is this TODO resolved?

cacheung committed:
fix format

fix format

cacheung committed:
Incorporate review feedback

Incorporate review feedback

cacheung committed:
update to use "presistedJson" directly in JSON comparison

update to use "presistedJson" directly in JSON comparison as per review comments.

cacheung commented:
@timkimadobe Thanks for the review! I update the tests based on the review comments.

addb commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/IdentityAdIdTest.java:
```diff
@@ -21,9 +27,11 @@
 import com.adobe.marketing.mobile.EventType;
 import com.adobe.marketing.mobile.MobileCore;
-import com.adobe.marketing.mobile.edge.identity.util.MonitorExtension;
+import com.adobe.marketing.mobile.util.JSONAsserts;
```
Nit: can we have all the util imports together

cacheung committed:
remove registerExtension in FunctionalTestUtil.

remove registerExtension in FunctionalTestUtil. It has been moved to testUtils.

timkimadobe commented on code/edgeidentity/src/androidTest/java/com/adobe/marketing/mobile/edge/identity/util/IdentityFunctionalTestUtil.java:
```diff
@@ -386,17 +312,32 @@ public static void verifyPrimaryECID(final String primaryECID) throws Exception
+			JSONAsserts.assertTypeMatch(json, xdmSharedState, new CollectionEqualCount("identityMap.ECID"));
+			return;
+		}
+		String expected =
```
Sorry I meant to comment on this JSON; right now it's validating the **first** element of the ECID array - we should update this to add a blank first element, so that the `secondaryECID` value is aligned in the second position:

```java
String expected =
    "{" +
    "  \"identityMap\": {" +
    "    \"ECID\": [" +
    "      {}," +  // Empty first ECID object
    "      {" +
    "        \"id\": \"" + secondaryECID + "\"" +
    "      }" +
    "    ]" +
    "  }" +
    "}";
```

timkimadobe reviewed (commented):
Thank you for the updates @cacheung! Just a few more follow up points

cacheung committed:
Update with review comments

Update with review comments
